services/rpcfilters/logs_cache.go

```diff
@@ -0,0 +1,132 @@
+package rpcfilters
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const (
+	cacheSize = 20
+)
+
+type cacheRecord struct {
+	block uint64
+	hash  common.Hash
+	logs  []types.Log
+}
+
+func newCache(size int) *cache {
+	return &cache{
+		records: make([]cacheRecord, size),
+		size:    size,
+	}
+}
+
+type cache struct {
+	mu      sync.RWMutex
+	last    int
+	size    int
+	records []cacheRecord
+}
+
+func (c *cache) add(logs []types.Log) (added, replaced []types.Log, err error) {
+	if len(logs) == 0 {
+		return nil, nil, nil
+	}
+	aggregated := aggregateLogs(logs, c.size) // size doesn't change
+	if len(aggregated) == 0 {
+		return nil, nil, nil
+	}
+	for prev, i := 0, 1; i < len(aggregated); i++ {
+		if aggregated[prev].block == aggregated[i].block-1 {
+			prev = i
+		} else {
+			return nil, nil, fmt.Errorf(
+				"logs must be delivered straight in order. gaps between blocks '%d' and '%d'",
+				aggregated[prev].block, aggregated[i].block,
+			)
+		}
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// find common block. e.g. [3,4] and [1,2,3,4] = 3
+	last := c.last
+	for aggregated[0].block < c.records[last].block && last > 0 {
+		last--
+	}
+	c.records, c.last, added, replaced = merge(last, c.records, aggregated)
+	return added, replaced, nil
+}
+
+func (c *cache) earliestBlockNum() uint64 {
+	if len(c.records) < 0 {
```

Comment:
Not possible to be lower than zero. 
```suggestion
	if len(c.records) == 0 {
```